examples/semver.peggy

```diff
@@ -0,0 +1,60 @@
+/**
+ * SemVer.org v2
+ * https://semver.org/spec/v2.0.0.html
+ * For unit tests see: https://github.com/dselman/peggy-semver
+ */
+semver
+  = versionCore:versionCore
+    pre:('-' @preRelease)?
+    build:('+' @build)?
+  {
+    return { versionCore, pre, build };
+  }
+
+versionCore
+  = major:$numericIdentifier '.' minor:$numericIdentifier '.' patch:$numericIdentifier
+  {
+    return {
+      major: parseInt(major, 10),
+      minor: parseInt(minor, 10),
+      patch: parseInt(patch, 10),
+    };
+  }
+
+preRelease
+  = head:$preReleaseIdentifier tail:('.' @$preReleaseIdentifier)*
+  {
+    return [head, ...tail];
+  }
+
+build
+  = head:$buildIdentifier tail:('.' @$buildIdentifier)*
+  {
+    return [head, ...tail];
+  }
+
+preReleaseIdentifier
+  = alphanumericIdentifier
+  / numericIdentifier
+
+buildIdentifier
+  = alphanumericIdentifier
+  / digit+
+
+alphanumericIdentifier
```

Comment:
According to spec, it should be `identifierChar* nonDigit identifierChar*`.

If you substitute everything into `alphanumericIdentifier`, it turns out the only reason why it exists is to avoid parse conflict in `preReleaseIdentifier`. If there is no `nonDigit`, it might match same things as `numericIdentifier`.

In PEG parsers there is no parse conflicts.

Reply:
I think this is all you need to comply to the spec:
```
semver = core:versionCore prerelease:("-" @prerelease)? build:("+" @build)? {
	return {...core, prerelease, build};
}
versionCore = major:numericId "." minor:numericId "." patch:numericId { return {major, minor, patch}; }
prerelease = head:prereleaseId tail:("." @prereleaseId)* { return [head, ...tail]; }
build = head:buildId tail:("." @buildId)* { return [head, ...tail]; }
prereleaseId = alnumId / numericId
buildId = alnumId / $[0-9]+
numericId = $("0" / [1-9][0-9]*)
alnumId = $[0-9a-z-]i+
```

Reply:
Actually `buildId` is just `alnumId`. It will never hit the second branch.

Reply:
> I think this is all you need to comply to the spec:
> 
> ```
> semver = major:numericId "." minor:numericId "." patch:numericId prerelease:("-" @prerelease)? build:("+" @build)? {
> 	return {major, minor, patch, prerelease, build};
> }
> prerelease = head:prereleaseId tail:("." @prereleaseId)* { return [head, ...tail]; }
> build = head:alnumId tail:("." @alnumId)* { return [head, ...tail]; }
> prereleaseId = alnumId / $("0" / [1-9][0-9]*)
> alnumId = $[0-9a-z-]i+
> ```

rule `numericId` is missing, maybe like so:

```pegjs
semver = major:numericId "." minor:numericId "." patch:numericId prerelease:("-" @prerelease)? build:("+" @build)? {
	return {major, minor, patch, prerelease, build};
}
prerelease = head:prereleaseId tail:("." @prereleaseId)* { return [head, ...tail]; }
build = head:alnumId tail:("." @alnumId)* { return [head, ...tail]; }
prereleaseId = alnumId / numericId 
alnumId = $[0-9a-z-]i+
numericId = $("0" / [1-9][0-9]*)
```

However, while this will successfully parse all [valid examples](https://github.com/dselman/peggy-semver/blob/main/test/semver-valid.txt) it will also successfully parse [invalid examples](https://github.com/dselman/peggy-semver/blob/main/test/semver-invalid.txt) (`1.2.3-0123`, `1.2.3-0123.0123`) but "_numeric identifiers MUST NOT include leading zeroes_" in the pre-release part.